ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ContractsTable.scala

```diff
@@ -0,0 +1,122 @@
+
+  override final def lookupContractKeyGlobally(
+      key: Key
+  )(implicit connection: Connection): Option[ContractId] =
+    SQL"""
+  WITH last_contract_key_create AS (
+         SELECT participant_events.*
+           FROM participant_events, parameters
+          WHERE event_kind = 10 -- create
+            AND create_key_hash = ${key.hash}
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          ORDER BY event_sequential_id DESC
+          LIMIT 1
+       )
+  SELECT contract_id
+    FROM last_contract_key_create -- creation only, as divulged contracts cannot be fetched by key
+   WHERE NOT EXISTS
+         (SELECT 1
+            FROM participant_events, parameters
+           WHERE event_kind = 20 -- consuming exercise
+             AND event_sequential_id <= parameters.ledger_end_sequential_id
+             AND contract_id = last_contract_key_create.contract_id
+         );
+       """
+      .as(contractId("contract_id").singleOpt)
+
+  override final def lookupMaximumLedgerTime(
+      ids: Set[ContractId]
+  )(implicit connection: Connection): Try[Option[Instant]] = {
+    if (ids.isEmpty) {
+      Failure(ContractsTable.emptyContractIds)
+    } else {
+      def lookup(id: ContractId): Option[Option[Instant]] =
+        SQL"""
+  WITH archival_event AS (
+         SELECT participant_events.*
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 20  -- consuming exercise
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          LIMIT 1
+       ),
+       create_event AS (
+         SELECT ledger_effective_time
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 10  -- create
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          LIMIT 1 -- limit here to guide planner wrt expected number of results
+       ),
+       divulged_contract AS (
+         SELECT ledger_effective_time
+           FROM participant_events, parameters
+          WHERE contract_id = $id
+            AND event_kind = 0 -- divulgence
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          ORDER BY event_sequential_id
+            -- prudent engineering: make results more stable by preferring earlier divulgence events
+            -- Results might still change due to pruning.
+          LIMIT 1
+       ),
+       create_and_divulged_contracts AS (
+         (SELECT * FROM create_event)   -- prefer create over divulgance events
+         UNION ALL
+         (SELECT * FROM divulged_contract)
+       )
+  SELECT ledger_effective_time
+    FROM create_and_divulged_contracts
+   WHERE NOT EXISTS (SELECT 1 FROM archival_event)
+   LIMIT 1;
+               """.as(instant("ledger_effective_time").?.singleOpt)
+
+      val foundIds: List[Option[Instant]] = ids.toList
```

Comment:
`foundIds` contains ledger effective times and not ids. 
```suggestion
      val foundLedgerEffectiveTimes: List[Option[Instant]] = ids.toList
```

Reply:
good idea, but this picture of the optionally found optional ledger effective time is a though for nomenclature. Wouldnt be missleading if we say here foundLedgerEffectiveTimes, when in fact those could be none? Also precise naming would be very chatty here

Reply:
changed